subsys/bluetooth/host/l2cap.c

```diff
@@ -451,15 +461,57 @@ static int l2cap_le_conn_req(struct bt_l2cap_le_chan *ch)
 	req->mps = sys_cpu_to_le16(ch->rx.mps);
 	req->credits = sys_cpu_to_le16(ch->rx.init_credits);
 
-	l2cap_chan_send_req(ch, buf, L2CAP_CONN_TIMEOUT);
+	l2cap_chan_send_req(&ch->chan, buf, L2CAP_CONN_TIMEOUT);
+
+	return 0;
+}
+
+static int l2cap_ecred_conn_req(struct bt_l2cap_chan **chan, int channels)
+{
+	struct net_buf *buf;
+	struct bt_l2cap_ecred_conn_req *req;
+	struct bt_l2cap_le_chan *ch;
+	int i;
+	u8_t ident;
+
+	if (!chan || !channels) {
+		return -EINVAL;
+	}
+
+	ident = get_ident();
+
+	buf = l2cap_create_le_sig_pdu(NULL, BT_L2CAP_ECRED_CONN_REQ, ident,
+				      sizeof(*req) +
+				      (channels * sizeof(u16_t)));
+
+	req = net_buf_add(buf, sizeof(*req));
+
+	ch = BT_L2CAP_LE_CHAN(chan[0]);
+
+	/* Init common parameters */
+	req->psm = sys_cpu_to_le16(ch->chan.psm);
+	req->mtu = sys_cpu_to_le16(ch->rx.mtu);
+	req->mps = sys_cpu_to_le16(ch->rx.mps);
+	req->credits = sys_cpu_to_le16(ch->rx.init_credits);
+
+	for (i = 0; i < channels; i++) {
+		ch = BT_L2CAP_LE_CHAN(chan[i]);
+
+		ch->chan.ident = ident;
+
+		net_buf_add_le16(buf, ch->rx.cid);
+	}
+
+	l2cap_chan_send_req(*chan, buf, L2CAP_CONN_TIMEOUT);
 
 	return 0;
 }
 
 static void l2cap_le_encrypt_change(struct bt_l2cap_chan *chan, u8_t status)
 {
-	/* Skip channels already connected or with a pending request */
-	if (chan->state != BT_L2CAP_CONNECT || chan->ident) {
+	/* Skip channels already connected or there is an ongoing request */
+	if (chan->state != BT_L2CAP_CONNECT ||
+	    k_delayed_work_remaining_get(&chan->rtx_work)) {
```

Comment:
Ive added a check for k_work_pending as well it the timer may have expired but the work has not been executed yet, note though that this is not canceling the work it is just checking if there a pending request which means the channel is not waiting for a encryption change to retry connecting.

Reply:
I think that could still be problematic. Since the timeout might have been expired, but not yet handled.
I tried to create helper function to address this issue, which should handle the case where the remaining value is zero. But @pabigot still didn't think this was safe enough:
https://github.com/zephyrproject-rtos/zephyr/pull/22826/files#diff-ed0b4d3815d3ce00b451fa30ed3c2051R143

In the Mesh similar problem is solved through state-flags instead.

Reply:
A better summary of my position is https://github.com/zephyrproject-rtos/zephyr/pull/22826#pullrequestreview-371142029.  I originally blocked that PR due to lack of a test; it seems to have a test, so I've removed my NAK.  Until there's a more robust `k_work` implementation I also an using external state to track at a gross level whether a submitted request is still pending.

Reply:
That sounds a little too paranoid here given that there is a state already, but I will go ahead and create a flag as well.